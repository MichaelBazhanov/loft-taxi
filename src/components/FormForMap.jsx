import React, { useState, useEffect } from "react";
import { connect } from 'react-redux'
import { getAddressList } from '../actions' //просто импортируем action
import Select from '../components/Select'

//img
import imgStandart from '../assets/images/car-standart.jpg'
import imgPremium from '../assets/images/car-premium.jpg'
import imgBusiness from '../assets/images/car-business.jpg'

//Car
import CarForForm from './CarForForm'

function classNames(...classes) {
	return classes.filter(Boolean).join(' ')
}

const FormForMap = ({ getAddressList, address, address1, address2 }) => {
	const [active, setActive] = useState(false)
	const [activeIndexCar, setActiveIndexCar] = useState(1)
	const [addressUSE, setAddressUSE] = useState([])

	const handleSubmit = (event) => {
		event.preventDefault();
		// делаем что то с данными
		alert(`Форма отправлена! ${active}`)
	}

	useEffect(() => {
		getAddressList()
	}, [])

	//================================================================= ЭТО занимает какое то время и до этого компонент не отображаем
	// console.log('рендер компонента :', address)
	useEffect(() => { // дополнительная обработка адресов
		let addr = address.map((el, inx) => {
			return { id: inx + 1, rout: el }
		})
		// console.log('useEffect :', addr)
		setAddressUSE(addr) // Установка useState запускает перерэндер компонента
	}, [address, address1, address2])
	// console.log('Установленный state in useEffect: ', addressUSE)
	//================================================================= ЭТО занимает какое то время и до этого компонент не отображаем



	return (
		<div className="container mx-auto h-screen relative pointer-events-none">

			<div className="flex flex-col">

				<form onSubmit={handleSubmit} className={classNames(active ? 'hidden' : '', 'max-w-[486px] w-full bg-white  mt-16 ml-24 rounded-xl shadow-lg pointer-events-auto')}>

					<div className="p-6 pb-0">
						{/* {addressUSE.length > 0 && <Select addressList={addressUSE} currentAddress={addressUSE[0]} idx='1' />}
						{addressUSE.length > 0 && <Select addressList={addressUSE} currentAddress={addressUSE[address.length - 1]} idx='2' />} */}

						{/* {address1 && <h2> address1 : {address1.rout}</h2>}
						{addressUSE && <h2>addressUSE : {addressUSE.map(item => item.rout)}</h2>} */}

						{/* {address1 && addressUSE.length > 0 && <Select addressList={addressUSE.filter(item => item.rout !== address1.rout)} currentAddress={addressUSE[0]} idx='1' />} */}
						{addressUSE.length > 0 && <Select addressList={addressUSE.filter(item => address1 ? item.rout !== address1.rout : true)} currentAddress={addressUSE[0]} idx='1' />}
						{addressUSE.length > 0 && <Select addressList={addressUSE.filter(item => address2 ? item.rout !== address1.rout : true)} currentAddress={addressUSE[address.length - 1]} idx='2' />}
					</div>

					<hr className="border w-full" />

					<div className="py-8 px-10 mt-6 rounded-xl border">
						<div className="flex -mx-3">
							<CarForForm price={'150 ₽'} imgSRC={imgStandart} index={1} setActiveIndexCar={setActiveIndexCar} activeIndexCar={activeIndexCar} />
							<CarForForm price={'200 ₽'} imgSRC={imgPremium} index={2} setActiveIndexCar={setActiveIndexCar} activeIndexCar={activeIndexCar} />
							<CarForForm price={'300 ₽'} imgSRC={imgBusiness} index={3} setActiveIndexCar={setActiveIndexCar} activeIndexCar={activeIndexCar} />
						</div>
						{/* верхний блок что то должен вернуть и я запишу это в инпуты */}
						<input type="hidden" name="car" />
						<button onClick={() => { setActive(true) }} type="submit" className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7">Заказать</button>
					</div>
				</form>

				<div className={classNames(active ? '' : 'hidden', 'max-w-[486px] w-full bg-white  mt-16 ml-24 rounded-xl shadow-lg py-10 px-11 pointer-events-auto')}>
					<p className="font-bold text-4xl">Заказ размещен</p>
					<p className="mt-4 text-lg text-gray-me">Ваше такси уже едет к вам. Прибудет приблизительно через 10 минут.</p>
					<button onClick={() => { setActive(false) }} type="submit" className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7">Сделать новый заказ</button>
				</div>

			</div>
		</div>
	)
}
export default connect(
	state => ({
		address: state.address.address,
		address1: state.routes.address1,
		address2: state.routes.address2
	}),
	{ getAddressList } // просто дергаем ACTION
)(FormForMap)