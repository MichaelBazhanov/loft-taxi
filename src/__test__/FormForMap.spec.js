import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import FormForMap from '../components/FormForMap'

// Мокаем useNavigate что бы внутри <FormForMap /> отдал просто функцию
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => jest.fn(),
}))
// Формируем замоканый стор
const mockStore = {
  getState: () => ({
    // входящие пропсы
    authorizationReducer: { token: '123' },

    addressReducer: {
      address: [
        { id: 1, root: 'Адрес 1' },
        { id: 2, root: 'Адрес 2' },
      ],
      isLoading: false,
      error: false,
    },
    paymentReducer: {
      cardName: 'test',
      cardNumber: '1',
      expiryDate: '000',
      cvc: '1',
      isLoadingSendPaymentCardNewUser: false,
      errorSendPaymentCardNewUser: false,
    },
  }),
  subscribe: () => {},
  dispatch: () => {
    // диспачим внутри компонента
    getAddressList: jest.fn()
    getRoutesCoordinates: jest.fn()
    getPaymentCard: jest.fn()
  },
}

it('Renders correctly', () => {
  const FormForMapExample = TestRenderer.create(
    <Provider store={mockStore}>
      <FormForMap />
    </Provider>,
  ).toJSON()

  expect(FormForMapExample).toMatchSnapshot()
})
