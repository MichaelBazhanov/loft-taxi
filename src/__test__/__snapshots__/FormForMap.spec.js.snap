// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  className="container mx-auto relative pointer-events-none"
>
  <div
    className="flex flex-col"
  >
    <form
      className="max-w-[486px] w-full bg-white  mt-16 ml-24 rounded-xl shadow-lg pointer-events-auto"
      onSubmit={[Function]}
    >
      <div
        className="p-6 pb-0"
      >
        <div
          className="mt-1 relative"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="relative w-full bg-white border-b border-gray-300 pl-3 pr-10 py-2 text-left cursor-default focus:outline-none focus:ring-1 focus:ring-yellow-me focus:border-yellow-me sm:text-sm"
            disabled={false}
            id="headlessui-listbox-button-1"
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            placeholder="Откуда"
            type="button"
          >
            <span
              className="flex items-center"
            >
              <div
                className="flex-shrink-0 h-3 w-3 rounded-full bg-black"
              />
              <span
                className="ml-3 block truncate"
              >
                Откуда
              </span>
            </span>
            <span
              className="ml-3 absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none divide-x-2"
            >
              <svg
                aria-hidden="true"
                className="h-5 w-5 text-gray-400 mr-1"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                  fillRule="evenodd"
                />
              </svg>
              <svg
                aria-hidden="true"
                className="h-8 w-8 text-gray-400"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          className="mt-1 relative"
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="relative w-full bg-white border-b border-gray-300 pl-3 pr-10 py-2 text-left cursor-default focus:outline-none focus:ring-1 focus:ring-yellow-me focus:border-yellow-me sm:text-sm"
            disabled={false}
            id="headlessui-listbox-button-2"
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            placeholder="Куда"
            type="button"
          >
            <span
              className="flex items-center"
            >
              <div
                className="flex-shrink-0 h-3 w-3 rounded-full bg-black"
              />
              <span
                className="ml-3 block truncate"
              >
                Куда
              </span>
            </span>
            <span
              className="ml-3 absolute inset-y-0 right-0 flex items-center pr-2 pointer-events-none divide-x-2"
            >
              <svg
                aria-hidden="true"
                className="h-5 w-5 text-gray-400 mr-1"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                  fillRule="evenodd"
                />
              </svg>
              <svg
                aria-hidden="true"
                className="h-8 w-8 text-gray-400"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
      <hr
        className="border w-full"
      />
      <div
        className="py-8 px-10 mt-6 rounded-xl border"
      >
        <div
          className="flex -mx-3"
        >
          <div
            className="w-1/3 px-3 cursor-pointer"
            onClick={[Function]}
          >
            <div
              className="flex flex-col shadow-me rounded-lg p-3"
            >
              <p>
                Стандарт
              </p>
              <p
                className="text-xs"
              >
                Стоимость
              </p>
              <p
                className="text-2xl"
              >
                150 ₽
              </p>
              <img
                alt="car"
                className="w-full h-auto"
                src="car-standart.jpg"
              />
            </div>
          </div>
          <div
            className="opacity-50 w-1/3 px-3 cursor-pointer"
            onClick={[Function]}
          >
            <div
              className="flex flex-col shadow-me rounded-lg p-3"
            >
              <p>
                Стандарт
              </p>
              <p
                className="text-xs"
              >
                Стоимость
              </p>
              <p
                className="text-2xl"
              >
                200 ₽
              </p>
              <img
                alt="car"
                className="w-full h-auto"
                src="car-premium.jpg"
              />
            </div>
          </div>
          <div
            className="opacity-50 w-1/3 px-3 cursor-pointer"
            onClick={[Function]}
          >
            <div
              className="flex flex-col shadow-me rounded-lg p-3"
            >
              <p>
                Стандарт
              </p>
              <p
                className="text-xs"
              >
                Стоимость
              </p>
              <p
                className="text-2xl"
              >
                300 ₽
              </p>
              <img
                alt="car"
                className="w-full h-auto"
                src="car-business.jpg"
              />
            </div>
          </div>
        </div>
        <input
          name="car"
          type="hidden"
        />
        <button
          className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7 disabled:opacity-75"
          disabled={true}
          onClick={[Function]}
          type="submit"
        >
          Заказать
        </button>
      </div>
    </form>
    <div
      className="hidden max-w-[486px] w-full bg-white  mt-16 ml-24 rounded-xl shadow-lg py-10 px-11 pointer-events-auto"
    >
      <p
        className="font-bold text-4xl"
      >
        Заказ размещен
      </p>
      <p
        className="mt-4 text-lg text-gray-me"
      >
        Ваше такси уже едет к вам. Прибудет приблизительно через 10 минут.
      </p>
      <button
        className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7"
        onClick={[Function]}
        type="submit"
      >
        Сделать новый заказ
      </button>
    </div>
    <div
      className="hidden max-w-[486px] w-full bg-white  mt-16 ml-24 rounded-xl shadow-lg py-10 px-11 pointer-events-auto"
    >
      <p
        className="font-bold text-4xl"
      >
        Заполните платежные данные
      </p>
      <p
        className="mt-4 text-lg text-gray-me"
      >
        Укажите информацию о платежной карте что бы сделать заказ.
      </p>
      <button
        className="text-2xl py-4 w-full bg-yellow-me rounded-full mt-7"
        onClick={[Function]}
        type="submit"
      >
        Перейти в профиль
      </button>
    </div>
  </div>
</div>
`;
